### ISSUE TITLE
Section_2.6_3.5_Figure3.py

### ISSUE THREAD

**Issue by Curiosity-Mars** [Curiosity-Mars/rarp-psa-trajectory-reproducibility#6]
"""Figure 3. Out-of-fold discrimination of PSA indicator models.

Manuscript alignment
--------------------
Methods: Section 2.6, Complementary Analysis of Linear and Nonlinear Approaches
Results: Section 3.5, Complementary Evaluation of Linear and Nonlinear Approaches

Prerequisites: run 00_Common_Setup.py, Table 2, and Figure 2.  The target is the
fixed computational Cluster 1 versus Clusters 0 and 2; it is not a prospective
clinical outcome.
"""

from __future__ import annotations

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import roc_auc_score, roc_curve

required_globals = (
    "OUTPUT_DIR",
    "TRAJECTORY_COHORT_N",
    "RANDOM_STATE",
    "N_BOOTSTRAP",
    "N_CV_SPLITS",
)
missing_globals = [name for name in required_globals if name not in globals()]
if missing_globals:
    raise RuntimeError(
        "Run 00_Common_Setup.py before this script. Missing: "
        + ", ".join(missing_globals)
    )

LINEAR_FILE = OUTPUT_DIR / "Table2" / "Table2_Per_Patient_Linear_Descriptors.csv"
NONLINEAR_FILE = OUTPUT_DIR / "Figure2" / "Figure2_Per_Patient_Nonlinear_Indicators.csv"
for path, prerequisite in (
    (LINEAR_FILE, "Section_2.4_3.3_Table2.py"),
    (NONLINEAR_FILE, "Section_2.5_3.4_Figure2.py"),
):
    if not path.exists():
        raise FileNotFoundError(f"Missing {path}. Run {prerequisite} first.")

OUTDIR = OUTPUT_DIR / "Figure3"
OUTDIR.mkdir(parents=True, exist_ok=True)

# -----------------------------------------------------------------------------
# Modeling table
# -----------------------------------------------------------------------------
linear = pd.read_csv(LINEAR_FILE)
nonlinear = pd.read_csv(NONLINEAR_FILE)
model_data = linear.merge(
    nonlinear,
    on=["Analysis_Row_ID", "Cluster"],
    how="inner",
    validate="one_to_one",
)
if len(model_data) != TRAJECTORY_COHORT_N or model_data["Analysis_Row_ID"].duplicated().any():
    raise AssertionError("Expected 1,434 unique patient records.")

def parse_boolean(series: pd.Series, allow_missing: bool = False) -> pd.Series:
    mapping = {
        True: True,
        False: False,
        "True": True,
        "False": False,
        "true": True,
        "false": False,
        1: True,
        0: False,
        1.0: True,
        0.0: False,
    }
    converted = series.map(mapping)
    if not allow_missing and converted.isna().any():
        raise ValueError(f"Unrecognized Boolean values: {series[converted.isna()].unique()}")
    return converted.astype("boolean")

model_data["indicator_evaluable"] = parse_boolean(model_data["indicator_evaluable"])
model_data["t_star_boundary"] = parse_boolean(
    model_data["t_star_boundary"], allow_missing=True
)

LINEAR_FEATURES = [
    "overall_raw_slope",
    "overall_log_slope",
    "raw_slope_POM0_24",
    "raw_slope_POM24_60",
    "raw_slope_POM60_120",
]
NONLINEAR_FEATURES = ["operational_t_star", "max_a_t", "mean_a_t"]
MODEL_FEATURES = {
    "Linear": LINEAR_FEATURES,
    "Nonlinear": NONLINEAR_FEATURES,
    "Integrative": LINEAR_FEATURES + NONLINEAR_FEATURES,
}

model_data["target_cluster1"] = (model_data["Cluster"] == 1).astype(int)
y = model_data["target_cluster1"].to_numpy(int)
if int(y.sum()) != 984 or int((y == 0).sum()) != 450:
    raise AssertionError("Unexpected target-class counts.")

def make_model() -> Pipeline:
    return Pipeline(
        [
            ("imputer", SimpleImputer(strategy="median")),
            ("scaler", StandardScaler()),
            (
                "classifier",
                LogisticRegression(
                    solver="lbfgs", max_iter=5000, random_state=RANDOM_STATE
                ),
            ),
        ]
    )

def fixed_splits(target: np.ndarray) -> list[tuple[np.ndarray, np.ndarray]]:
    cv = StratifiedKFold(
        n_splits=N_CV_SPLITS, shuffle=True, random_state=RANDOM_STATE
    )
    return list(cv.split(np.zeros(len(target)), target))

def oof_probabilities(
    data: pd.DataFrame,
    target: np.ndarray,
    feature_sets: dict[str, list[str]],
) -> tuple[pd.DataFrame, pd.DataFrame]:
    splits = fixed_splits(target)
    output = pd.DataFrame(index=data.index)
    output["fold"] = -1
    fold_rows = []
    for fold, (_, test_index) in enumerate(splits, start=1):
        output.iloc[test_index, output.columns.get_loc("fold")] = fold

    for name, features in feature_sets.items():
        probability = np.full(len(data), np.nan)
        for fold, (train_index, test_index) in enumerate(splits, start=1):
            model = make_model()
            model.fit(data.iloc[train_index][features], target[train_index])
            probability[test_index] = model.predict_proba(
                data.iloc[test_index][features]
            )[:, 1]
            fold_rows.append(
                {
                    "Model": name,
                    "Fold": fold,
                    "N_test": len(test_index),
                    "ROC_AUC": roc_auc_score(target[test_index], probability[test_index]),
                }
            )
        output[f"probability_{name}"] = probability
    if output.isna().any().any() or (output["fold"] < 1).any():
        raise AssertionError("Incomplete out-of-fold prediction table.")
    return output, pd.DataFrame(fold_rows)

def bootstrap_auc(
    target: np.ndarray, probability: np.ndarray, seed: int = RANDOM_STATE
) -> np.ndarray:
    rng = np.random.default_rng(seed)
    values = []
    for _ in range(N_BOOTSTRAP):
        index = rng.integers(0, len(target), len(target))
        if np.unique(target[index]).size == 2:
            values.append(roc_auc_score(target[index], probability[index]))
    return np.asarray(values, float)

def paired_auc_difference(
    target: np.ndarray,
    first: np.ndarray,
    second: np.ndarray,
    seed: int = RANDOM_STATE,
) -> tuple[float, float, float, float]:
    rng = np.random.default_rng(seed)
    differences = []
    for _ in range(N_BOOTSTRAP):
        index = rng.integers(0, len(target), len(target))
        if np.unique(target[index]).size == 2:
            differences.append(
                roc_auc_score(target[index], first[index])
                - roc_auc_score(target[index], second[index])
            )
    differences = np.asarray(differences, float)
    p_value = 2 * min(np.mean(differences <= 0), np.mean(differences >= 0))
    return (
        float(np.mean(differences)),
        float(np.percentile(differences, 2.5)),
        float(np.percentile(differences, 97.5)),
        min(float(p_value), 1.0),
    )

# -----------------------------------------------------------------------------
# Primary out-of-fold analysis
# -----------------------------------------------------------------------------
oof_prob, fold_metrics = oof_probabilities(model_data, y, MODEL_FEATURES)
oof = model_data[["Analysis_Row_ID", "Cluster", "target_cluster1"]].copy()
oof = pd.concat([oof, oof_prob], axis=1)

performance_rows = []
for index, name in enumerate(MODEL_FEATURES):
    probability = oof[f"probability_{name}"].to_numpy(float)
    auc = float(roc_auc_score(y, probability))
    boot = bootstrap_auc(y, probability, RANDOM_STATE)
    performance_rows.append(
        {
            "Model": name,
            "N": len(y),
            "ROC_AUC": auc,
            "CI_low": float(np.percentile(boot, 2.5)),
            "CI_high": float(np.percentile(boot, 97.5)),
        }
    )
performance = pd.DataFrame(performance_rows)

comparison_rows = []
for first, second in (
    ("Nonlinear", "Linear"),
    ("Integrative", "Linear"),
    ("Integrative", "Nonlinear"),
):
    difference = paired_auc_difference(
        y,
        oof[f"probability_{first}"].to_numpy(float),
        oof[f"probability_{second}"].to_numpy(float),
    )
    comparison_rows.append(
        {
            "Comparison": f"{first} − {second}",
            "N": len(y),
            "Delta_AUC": difference[0],
            "CI_low": difference[1],
            "CI_high": difference[2],
            "P_bootstrap_two_sided": difference[3],
        }
    )
comparisons = pd.DataFrame(comparison_rows)

# -----------------------------------------------------------------------------
# Prespecified sensitivity analyses
# -----------------------------------------------------------------------------
def sensitivity_performance(data: pd.DataFrame, label: str) -> pd.DataFrame:
    target = data["target_cluster1"].to_numpy(int)
    probabilities, _ = oof_probabilities(data, target, MODEL_FEATURES)
    rows = []
    for name in MODEL_FEATURES:
        probability = probabilities[f"probability_{name}"].to_numpy(float)
        boot = bootstrap_auc(target, probability, RANDOM_STATE)
        rows.append(
            {
                "Analysis": label,
                "Model": name,
                "N": len(data),
                "ROC_AUC": roc_auc_score(target, probability),
                "CI_low": np.percentile(boot, 2.5),
                "CI_high": np.percentile(boot, 97.5),
            }
        )
    return pd.DataFrame(rows)

evaluable_data = model_data.loc[model_data["indicator_evaluable"].fillna(False)].reset_index(
    drop=True
)
evaluable_sensitivity = sensitivity_performance(evaluable_data, "Evaluable only")

boundary_data = model_data.copy()
boundary_flag = boundary_data["t_star_boundary"].fillna(False)
boundary_data.loc[boundary_flag, "operational_t_star"] = np.nan
boundary_sensitivity = sensitivity_performance(
    boundary_data, "Boundary t* treated as missing"
)
sensitivity = pd.concat(
    [evaluable_sensitivity, boundary_sensitivity], ignore_index=True
)

# -----------------------------------------------------------------------------
# Figure 3
# -----------------------------------------------------------------------------
fig, axes = plt.subplots(1, 3, figsize=(16, 5))
for name in MODEL_FEATURES:
    probability = oof[f"probability_{name}"].to_numpy(float)
    false_positive, true_positive, _ = roc_curve(y, probability)
    auc = roc_auc_score(y, probability)
    axes[0].plot(false_positive, true_positive, label=f"{name} (AUC={auc:.3f})")
axes[0].plot([0, 1], [0, 1], "--", linewidth=1, color="0.4")
axes[0].set_xlabel("False-positive rate")
axes[0].set_ylabel("True-positive rate")
axes[0].set_title("(a) Out-of-fold discrimination")
axes[0].legend(loc="lower right")

ordered = performance.set_index("Model").loc[list(MODEL_FEATURES)]
x = np.arange(len(ordered))
auc_values = ordered["ROC_AUC"].to_numpy(float)
error = np.vstack(
    [
        auc_values - ordered["CI_low"].to_numpy(float),
        ordered["CI_high"].to_numpy(float) - auc_values,
    ]
)
axes[1].errorbar(x, auc_values, yerr=error, fmt="o", capsize=4)
axes[1].set_xticks(x, ordered.index, rotation=20)
axes[1].set_ylabel("ROC AUC")
axes[1].set_ylim(0.45, 1.0)
axes[1].set_title("(b) Bootstrap 95% CI")
axes[1].axhline(0.5, linestyle="--", linewidth=1, color="0.4")

analysis_order = ("Evaluable only", "Boundary t* treated as missing")
for name in MODEL_FEATURES:
    subset = sensitivity.loc[sensitivity["Model"] == name].set_index("Analysis").loc[
        list(analysis_order)
    ]
    axes[2].plot(
        np.arange(2), subset["ROC_AUC"].to_numpy(float), marker="o", label=name
    )
axes[2].set_xticks(np.arange(2), ["Evaluable only", "Boundary t*\nas missing"])
axes[2].set_ylabel("ROC AUC")
axes[2].set_ylim(0.45, 1.0)
axes[2].set_title("(c) Sensitivity analyses")
axes[2].legend()

fig.suptitle(
    "Out-of-fold discrimination of linear, nonlinear, and integrative models",
    fontsize=14,
    y=1.02,
)
fig.tight_layout()
figure_png = OUTDIR / "Figure3_Model_Discrimination.png"
figure_pdf = OUTDIR / "Figure3_Model_Discrimination.pdf"
fig.savefig(figure_png, dpi=600, bbox_inches="tight", facecolor="white")
fig.savefig(figure_pdf, bbox_inches="tight", facecolor="white")
plt.close(fig)

# -----------------------------------------------------------------------------
# Published-value completion gate and outputs
# -----------------------------------------------------------------------------
expected_auc = {"Linear": 0.640324, "Nonlinear": 0.735257, "Integrative": 0.734077}
for model, expected in expected_auc.items():
    observed = float(performance.loc[performance["Model"] == model, "ROC_AUC"].iloc[0])
    if not np.isclose(observed, expected, atol=5e-7, rtol=0):
        raise AssertionError(f"{model} AUC: expected {expected}; found {observed}")

expected_sensitivity_auc = {
    ("Evaluable only", "Linear"): 0.635095,
    ("Evaluable only", "Nonlinear"): 0.735572,
    ("Evaluable only", "Integrative"): 0.759047,
    ("Boundary t* treated as missing", "Linear"): 0.640324,
    ("Boundary t* treated as missing", "Nonlinear"): 0.734767,
    ("Boundary t* treated as missing", "Integrative"): 0.733899,
}
for key, expected in expected_sensitivity_auc.items():
    analysis_name, model = key
    observed = float(
        sensitivity.loc[
            (sensitivity["Analysis"] == analysis_name) & (sensitivity["Model"] == model),
            "ROC_AUC",
        ].iloc[0]
    )
    if not np.isclose(observed, expected, atol=5e-7, rtol=0):
        raise AssertionError(f"{key} AUC: expected {expected}; found {observed}")

oof.to_csv(OUTDIR / "Figure3_OOF_Predictions.csv", index=False)
fold_metrics.to_csv(OUTDIR / "Figure3_Fold_Metrics.csv", index=False)
performance.to_csv(OUTDIR / "Figure3_Model_Performance.csv", index=False)
comparisons.to_csv(OUTDIR / "Figure3_Paired_AUC_Comparisons.csv", index=False)
sensitivity.to_csv(OUTDIR / "Figure3_Sensitivity_Analyses.csv", index=False)

print(performance.to_string(index=False))
print(comparisons.to_string(index=False))
print("\nFIGURE 3 COMPLETION GATE PASSED")
print("Figure:", figure_png)
